### PR TITLE
JDK-8296796: Provide clean, platform-agnostic interface to C-heap trimming

### PR DESCRIPTION
**Pull request by tstuefe**
This is a breakout from [JDK-8293114](https://bugs.openjdk.org/browse/JDK-8293114), which is starved for reviews. So I attempt to break up that fix into smaller units which are hopefully easier to review separately.

We can trim the C-heap manually using jcmd since [JDK-8268893](https://bugs.openjdk.org/browse/JDK-8268893). This patch reshapes this code, cleaning it up in an OS-agnostic way. That will allow us to add implementions for other platforms (I have this on my list for AIX at least) and make review of 8293114 easier.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296796](https://bugs.openjdk.org/browse/JDK-8296796): Provide clean, platform-agnostic interface to C-heap trimming

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [1a642e0d](https://git.openjdk.org/jdk/pull/11089/files/1a642e0d0a6a6ec55ef804d168cd8017b2157435)
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [1a642e0d](https://git.openjdk.org/jdk/pull/11089/files/1a642e0d0a6a6ec55ef804d168cd8017b2157435)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11089/head:pull/11089` \
`$ git checkout pull/11089`

Update a local copy of the PR: \
`$ git checkout pull/11089` \
`$ git pull https://git.openjdk.org/jdk pull/11089/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11089`

View PR using the GUI difftool: \
`$ git pr show -t 11089`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11089.diff">https://git.openjdk.org/jdk/pull/11089.diff</a>

</details>

### CODE REVIEW

**Comment by tstuefe** [comment]
x86 test error unrelated

File: src/hotspot/share/utilities/globalDefinitions.hpp [hunk 1]
```diff
@@ -372,6 +372,9 @@ inline T byte_size_in_proper_unit(T s) {
   }
 }
 
+#define PROPERFMT             SIZE_FORMAT "%s"
+#define PROPERFMTARGS(S)      byte_size_in_proper_unit(S), proper_unit_for_byte_size(S)
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/globalDefinitions.hpp:376]: style nit? lower-case 's'

**Review by dholmes-ora — COMMENTED** [review]
This looks good for doing what it says, but I have to wonder whether it is actually worthwhile doing this unless most OS/lib will support it? What will the implementation be in AIX?

Thanks

**Comment by tstuefe** [comment]
> This looks good for doing what it says, but I have to wonder whether it is actually worthwhile doing this unless most OS/lib will support it? What will the implementation be in AIX?

I think C-Heap trimming is useful even if only Linux does it. Linux is arguably the most important platform. And https://github.com/openjdk/jdk/pull/10085 would bring demonstratable benefits but did not garner a lot of interest. So I hope to speed it up by splitting parts that are hopefully non-contentious into separate RFEs.

I know that AIX has an API to disclaim memory, but have not yet looked deeply into the integration. I also cannot rule out that other platforms may give us similar APIs in the future.

**Comment by dholmes-ora** [comment]
I don't disagree that C-heap trimming is useful even if only Linux does it. My objection is to defining a platform-agnostic API when only Linux does it.

**Comment by tstuefe** [comment]
> I don't disagree that C-heap trimming is useful even if only Linux does it. My objection is to defining a platform-agnostic API when only Linux does it.

I see. It will allow us to use these APIs in shared code though, without having to use platform ifdefs. We do similar things for other platforms (e.g. see os::map_stack_shadow_pages).

**Review by dholmes-ora — APPROVED** [review]
Okay. I have some reservations about this style of approach but the precedents are there. I'd argue that for single-use situations like this and os::map_stack_shadow_pages that a XXX_ONLY(foo();) in the shared code would be acceptable. Others may disagree.

Thanks.

**Comment by tstuefe** [comment]
> Okay. I have some reservations about this style of approach but the precedents are there. I'd argue that for single-use situations like this and os::map_stack_shadow_pages that a XXX_ONLY(foo();) in the shared code would be acceptable. Others may disagree.
> 

Thanks David. I myself lean towards ifdefs too than having to search for the one platform implementation that actually does something. But I caught flak in the past for too many ifdefs. If it annoys us after [JDK-8293114](https://bugs.openjdk.org/browse/JDK-8293114) and no second platform implementation is forthcoming, I'll remove the APIs from shared os.

**Review by rkennke — APPROVED** [review]
Looks good to me. I kinda like the OS interface. If the proliferation of lots of unimplemented methods in os/* is a concern, we could provide default impls in shared if !__GLIBC__ as a middle-ground. WDYT?

**Comment by tstuefe** [comment]
> Looks good to me.

Thank you!

> I kinda like the OS interface. If the proliferation of lots of unimplemented methods in os/* is a concern, we could provide default impls in shared if !**GLIBC** as a middle-ground. WDYT?

I fear this would annoy people even more than platform defines. Ioi did use something with "HAVE_xxxx_xxxx" macros to provide default implementations for similar things, but I think in this case we are talking about so little code, its not worth the trouble.

**Comment by tstuefe** [comment]
Okay, I manually tested fastdebug on Alpine and on 32-bit. Also re-merged. If all tests run through with no attributable errors, I'll push.

Thanks @dholmes-ora @rkennke for reviewing!

**Comment by dholmes-ora** [comment]
We had a strange quirk in our CI testing after this was pushed and it made me think a little differently about the change. If `rss_change` is only needed for the test then we should perhaps make it (and the test) debug only? Also the `rss_change` logic seems not to be thread-safe, so only useful for that single-threaded test.

**Pull request closed without merging**